syntax = "proto3";
package market;


message Lot {
  string item_type = 1;
  string item_id = 2;
  uint32 price = 3;
  uint32 owner_id = 4;
}

message LotInfo {
  string item_type = 1;
  string item_id = 2;
  uint32 price = 3;
  double created_at = 4;
}


message ItemTypeSummary {
  string item_type = 1;
  uint32 sell_number = 2;
  uint32 min_sell_price = 3;
  uint32 max_sell_price = 4;
  uint32 owner_sell_number = 5;
}

message LogRecord {
  string item_type = 1;
  double created_at = 2;
  uint32 price = 3;
}


message PlaceSellLotRequest {
  repeated Lot lots = 1;
}


message PlaceSellLotResponse {
  repeated string lots_ids = 1;
}


message CloseSellLotRequest {
  string item_type = 1;
  uint32 price = 2;
  uint32 number = 3;
  uint32 buyer_id = 4;
}


message CloseSellLotResponse {
  repeated Lot lots = 1;
}


message CancelSellLotRequest {
  string item_type = 1;
  uint32 price = 2;
  uint32 number = 3;
  uint32 owner_id = 4;
}


message CancelSellLotResponse {
  repeated Lot lots = 1;
}

message CancelSellLotsByTypeRequest {
  string item_type = 1;
}


message CancelSellLotsByTypeResponse {
  repeated Lot lots = 1;
}

message ListSellLotsRequest {
  uint32 owner_id = 1;
}

message ListSellLotsResponse {
  repeated LotInfo lots = 1;
}

message InfoRequest {
  oneof owner {
    uint32 owner_id = 1;
  }
}


message InfoResponse {
  repeated ItemTypeSummary info = 1;
}


message ItemTypePricesRequest {
  string item_type = 1;
  oneof owner {
    uint32 owner_id = 2;
  }
}


message ItemTypePricesResponse {
  map<uint32, uint32> prices = 1;
  map<uint32, uint32> owner_prices = 2;
}


message HistoryRequest {
  uint32 page = 1;
  uint32 records_on_page = 2;
}


message HistoryResponse {
  uint32 total_records = 1;
  uint32 page = 2;
  repeated LogRecord records = 3;
}

message StatisticsRequest {
  double time_from = 1;
  double time_till = 2;
}


message StatisticsResponse {
  uint32 sell_lots_placed = 1;
  uint32 sell_lots_closed = 2;
  uint32 turnover = 3;
}


message DebugClearServiceRequest {}
message DebugClearServiceResponse {}
